11007 isPrime does not work for large numbers

johnhwoods opened the pull request:
`LargeInteger>>isPrime` overrides `Integer>>isPrime` with `^self isProbablyPrime` - a 25 iteration Miller/Rabin test.

However, because the fractional indices in the algorithm were expressed as floats, this stopped `isPrime` being useful above `Float fmax` (approx 1.8e308).  However, there's no performance reason for such a ceiling (and, indeed, many of the LargeIntegers we want to test for probable primality are cryptographic numbers in excess of 2^1024 and therefore larger than this).   The new method will return true, in a minute or so, on average 2020 PC, for the 23rd Mersenne Prime, 2^11213-1, a 3376 digit number.

johnhwoods committed:
1. Previously used floats as Miller/Rabin indices, leading to an effective ceiling of Float Maxval, approx 1.8e308.
fixed by replacing the 25 pregenerated floats with fractions.

2. Previously used  array do: [:e | ... ifFalse: [^false}]. ^true.
rewritten to use the slightly more elegant  ^array allSatisfy: [...]

johnhwoods committed:
Improvements to Integeger>> isProbablyPrime to allow receivers > Float fmax

1. Previously used floats as Miller/Rabin indices, leading to an ceiling of approx 1.8e308.
fixed by replacing the 25 pregenerated floats with fractions.

2. Previously used  array do: [:e | ... ifFalse: [^false}]. ^true.
rewritten to use the slightly more elegant  ^array allSatisfy: [...]

astares commented on src/Math-Operations-Extensions/Integer.extension.st:
```diff
@@ -146,29 +146,18 @@ Integer >> isPrime [
 Integer >> isProbablyPrime [
 	"See isProbablyPrimeWithK:andQ:randomIndex: for the algoritm description."
```
can you change typo: algorithm

johnhwoods replied:
Hah, missed that.  In my defence that typo is in the original method :-)

johnhwoods committed:
Fixed typo spotted by @astares

johnhwoods committed:
Merge branch 'pharo-project:Pharo10' into 11007-isPrime-does-not-work-for-large-numbers

MarcusDenker commented:
failig test not related:

- [CI] Failing randomly: Spec related tests with DNU SmallFloat64 >> #nextCell: #10522

Ducasse commented:
Does it fix #11046

Ducasse commented:
I'm confused what is the difference with PR about isPrime that got integrated

johnhwoods commented:
Hi, there were three separate issues which I idiotically turned into four.

1) isPrime didn't work for large numbers (where it redirects to isProbablyPrime which used a method which contained floats instead of fractions for its random indices in a Miller/Rabin method.  I changed the floats to fracs and it worked)  This was #11007, opened by Pavel.

2) I noticed when doing it that the SmallInteger isPrime had a 2x inefficiency so I raised #11018, and fixed this.

3) I also noticed primeFactors could be improved, so I raised #11045.  Whilst doing this I realised that

4) Unfortunately when I did (2) I introduced a stupid bug.  So I raised #11046.

Have I earned a ban yet?  :-D
